ramips: add support for Beeline SmartBox GIGA

csharper2005 opened the pull request:
[Device page on OpenWrt WiKi](https://openwrt.org/toh/beeline/smartbox_giga)
-----------------------------------------------------------------------------------------

Beeline SmartBox GIGA is a wireless WiFi 5 router manufactured by Sercomm and widely distributed by Beeline ISP.

Device specification
--------------------
SoC Type: MediaTek MT7621AT
RAM: 256 MiB, Nanya NT5CC128M16JR-EK
Flash: 128 MiB, Macronix MX30LF1G18AC
Wireless 2.4 GHz (MT7603EN): b/g/n, 2x2
Wireless 5 GHz (MT7613BE): a/n/ac, 2x2
Ethernet: 3 ports - 2xGbE (WAN, LAN1), 1xFE (LAN2)
USB ports: 1xUSB3.0
Button: 1 button (Reset/WPS)
PCB ID: DBE00B-1.6MM
LEDs: 1 RGB LED
Power: 12 VDC, 1.5 A
Connector type: barrel
Bootloader: U-Boot

Installation
-----------------
1. Downgrade stock (Beeline) firmware to v.1.0.02;
2. Give factory OpenWrt image a shorter name, e.g. 1001.img;
3. Upload and update the firmware via the original web interface.

Remark: You might need make the 3rd step twice if your running firmware is booted from the Slot 1 (Sercomm0 bootflag). The stock firmware reverses the bootflag (Sercomm0 / Sercomm1) on each firmware update.

Revert to stock
---------------
1. Change the bootflag to Sercomm1 in OpenWrt CLI and then reboot:
`      printf 1 | dd bs=1 seek=7 count=1 of=/dev/mtdblock3`
2. Update with any stock (Beeline) firmware to overwrite OpenWrt in Slot 0 completely.

MAC Addresses
-------------
```
+-----+-----------+---------+
| use | address   | example |
+-----+-----------+---------+
| LAN | label     | *:16    |
| WAN | label + 1 | *:17    |
| 2g  | label + 4 | *:1a    |
| 5g  | label + 5 | *:1b    |
+-----+-----------+---------+
```
The label MAC address was found in Factory 0x21000

Notes
-----
1. The following scripts are required for the build:
- sercomm-crypto.py - already exists in OpenWrt
- sercomm-partition-tag.py - already exists in OpenWrt
- sercomm-payload.py - already exists in OpenWrt
- sercomm-pid.py - new, the part of this pull request
- sercomm-kernel-header.py - new, the part of this pull request

2. This device (same as other Sercomm S2,S3-based devices) requires special LZMA and LOADADDR settings for successful boot:
- LZMA_TEXT_START=0x82800000
- KERNEL_LOADADDR=0x81001000
- LOADADDR=0x80001000

3. This device (same as several other Sercomm-based devices - Beeline, Netgear, Etisalat, Rostelecom) has partition map (mtd1) containing real partition offsets, which may differ from device to device depending on the number and location of bad blocks on NAND. "fixed-partitions" is used if the partition map is not found or corrupted. This behavour (it's the same as on stock firmware) is provided by MTD_SERCOMM_PARTS module.

Forum topic:
[Add support for Beeline SmartBox Giga on stock U-Boot](https://forum.openwrt.org/t/add-support-for-beeline-smartbox-giga-on-stock-u-boot/99390)

Related devices:
1. [Sercomm S3](https://forum.openwrt.org/t/add-support-for-sercomm-s3-on-stock-uboot/95970);
2. [Beeline Smartbox Turbo+](https://forum.openwrt.org/t/add-support-for-beeline-smartbox-turbo/99635), https://github.com/openwrt/openwrt/pull/4108

adschm commented:
Oh please, why this double-posting all the time?

adschm commented:
... and still with a commit title not fulfilling the rules in submitting-patches, now that you had another reminder to read it when opening the PR...

csharper2005 commented:
@adschm, 
I see. Sorry for that. But your comment has given me an insight. I've read the rules many times and assumed that they are related to a first post of the conversation at github (not to a commit message). All is fixed for now. What next steps are necessary from my side? 

P.S. I've looked at a lot of PR's: common mistakes, waste of resources (both requesters and maintainers). Let's add Top10 mistakes to a FAQ. And maybe few words about the OpenWRT change management process. How do you like this idea?

adschm commented:
Hmm, replacing bootloader is typically a no-go (at least on this target) ...

csharper2005 commented:
@adschm, 
Ramips devices typically have U-Boot with tftp boot enabled (activated with reset button) for emergency cases and stock fw recovery, but this one - does not. Easy to get brick. And... It's not a one way ticket. U-Boot can be restored if necessary from a backup.

csharper2005 commented:
@adschm, original firmware image (i.e. https://static.beeline.ru/upload/images/sbgiga2001.img) is encrypted. I've found this tool to decrypt it - https://github.com/Psychotropos/sercomm_fwutils. Have you ever faced factory image encryption for ramips devices? 
```
test@vmu1804:~/sercomm_fwutils$ python decrypt_image.py 2001.img 2001_decrypted.img
[+] Trying to interpret file as type1 image...
[+] Image key: 4204035580c88f71c3dcd7d6090329490c909a797e4ed146b5fe0ec50f95b3ae
[+] Image IV: 0f95ebc4f547486867bab830d7949012
[+] Wrote decrypted image to 2001_decrypted.img!
```

csharper2005 commented:
Additionally a special kernel header is required for the original bootloader... @kar200 has described the structure here  https://forum.openwrt.org/t/add-support-for-sercomm-s3-on-stock-uboot/95970/1  

One question while we are dealing with this. @adschm, why sorting is not necessary here - target/linux/ramips/image/mt7621.mk? e.g. IMAGES += is before DEVICE_VENDOR and so on.

kar200 commented:
> Additionally a special kernel header is required for the original bootloader... @kar200 has described the structure here https://forum.openwrt.org/t/add-support-for-sercomm-s3-on-stock-uboot/95970/1
> 
> One question while we are dealing with this. @adschm, why sorting is not necessary here - target/linux/ramips/image/mt7621.mk? e.g. IMAGES += is before DEVICE_VENDOR and so on.

Not in this case. This is replacing the bootloader so no neeed to add anything.

csharper2005 commented:
> Not in this case. This is replacing the bootloader so no neeed to add anything.

Right. But if I understood @adschm correctly, replacing bootloader is the stopper for this commit. Therefore, I am looking for an alternative option using the standard U-boot.

kar200 commented:
> Right. But if I understood @adschm correctly, replacing bootloader is the stopper for this commit. Therefore, I am looking for an alternative option using the standard U-boot.

My bad I just read the comment from @adschm. In this case I am sorry it is not as as simple as it using breed (for my level anyway).

csharper2005 commented:
> My bad I just read the comment from @adschm. In this case I am sorry it is not as as simple as it using breed (for my level anyway).

I've written a script that adds necessary header to a kernel (https://forum.openwrt.org/t/add-support-for-sercomm-s3-on-stock-uboot/95970/30?u=csharper2005). 

So, nowaday I can use U-Boot and follow the way described here https://github.com/openwrt/openwrt/pull/4019, but:
1. This requires opening the device case and soldering the UART (+UART adapter itself of course). Not very user friendly, is it? Сurrent method doesn't require an intervention in the hardware. 
2. Openwrt won't install on a devices with bad erase block(s) before the kernel partition. Details - https://github.com/openwrt/openwrt/pull/4019#issuecomment-860228204. I've checked that. Current method allows installation on such devices.  
3. New rootfs partition will be 24 bytes instead of the current 117 mbytes. lol 
In this regard, I have doubts that further work should be continued. I think that working product is more important than a stock bootloader.

kar200 commented:
hi @csharper2005 

As far as I understand it the openwrt philosphy is about having a firmware installed in the easiest way with the least changes to the device as well as an easy way to go back to stock. That's why they would not accept bootloader changes "I think"/

There is some work in here being done for the linksys devices. It already works in the mvebu platform. 
https://github.com/openwrt/openwrt/pull/3957

My understanding is that the parser is patched to check the uboot command line arguments. It then renames the partitions so the boot process can mount a different mtd (opposite to the static way it is done with UBI).

For the sercomm devices I don't think uboot provides the kernel with that nor does it use a nvram. It just reads the value (Sercomm0/1) and mounts the rootfs. From the bootlog it does not seem that any renaming happen and the partitions keep they names.

Maybe @arrmo would have an idea on how this can be achived by reading a value in the mtd itself. 



> * This requires opening the device case and soldering the UART (+UART adapter itself of course). Not very user friendly, is it? Сurrent method doesn't require an intervention in the hardware.

Some devices already use a workaround to this. It is by flashing the same firmware again which should take you to the opposite bootflag. I beleive this would be the best option as it would make going back to the stock firmware a breeze.

arrmo commented:
> Maybe @arrmo would have an idea on how this can be achived by reading a value in the mtd itself.
> 
> >

Sorry, long thread here - not sure of the question 😆

csharper2005 commented:
> > Maybe @arrmo would have an idea on how this can be achived by reading a value in the mtd itself.
> > >
> 
> Sorry, long thread here - not sure of the question 😆

Hi, @arrmo! What is the best way to read and change a hex value in raw mtd partition by a offset?
For example, mtd3, label "Boot Flag", offset 0x00001

arrmo commented:
> What is the best way to read and change a hex value in raw mtd partition by a offset?

Just thinking out loud here, but ... make sure the partition is r/w (not ro) => then store to a file (cat), edit (as binary), write back? I admit, I haven't really tried that specifically. I have messed with the u-boot env, but not just a random location. Sorry!

csharper2005 commented:
Can somebody help me with bad erase block handling U-Boot? The crux of the problem is described here -  https://forum.openwrt.org/t/add-support-for-beeline-smartbox-giga-on-stock-u-boot/99390/2?u=csharper2005.

csharper2005 commented:
FS#3887 - MTD partition offset not correctly mapped when bad eraseblocks present:
https://bugs.openwrt.org/index.php?do=details&task_id=3887

csharper2005 commented:
Changes (Breed -> U-Boot):
- Script that calculates the kernel header was added
- Layout was changed (based on a stock)
- Sercomm U-boot compatible loader compiling settings was added

csharper2005 commented:
> Hmm, replacing bootloader is typically a no-go (at least on this target) ...

@adschm, I have changed the pr. Stock U-boot is used instead of breed.

csharper2005 commented:
> Just thinking out loud here, but ... make sure the partition is r/w (not ro) => then store to a file (cat), edit (as binary), write back? I admit, I haven't really tried that specifically. I have messed with the u-boot env, but not just a random location. Sorry!

Fortunately, this Sercomm device has no boot counters. Newertheless, I found the appropriate command, that can be used for such cases:
`echo -ne "$(echo 31 | sed 's/../\\x&/g')" | dd of=/dev/mtdblock3 bs=1 seek=7 count=1`

kar200 commented:
Sorry @arrmo I tagged you here for a different reason (I got confused about the issue).

It was to do with your pull request for dual boot used in some mvebu device that you are trying to use with ramips.

Althoguh I don't understand completely I believe it changes the name of the active rootfs during boot depending on what the CMD line from uboot is.

On the sercomm devices (not sure about this one but definately for the turbo+) it uses a string in a bootflag partition (the only value there ) and on that it mounts the rootfs1 or rootfs2. There is no renaming of the MTD during (on the stock firmware that is).

I was wondering if there was a way to achieve this ?

csharper2005 commented:
> I was wondering if there was a way to achieve this ?
I squashed all partitions after the second kernel into the one big ubi. So there is no problem even the bootflag set up to Sercomm1. U-boot checks rootfs2, gets wrong CRC and switches to Sercomm0 with OpenWrt. On your device (and turbo+ too) there is a bootcounter in mtd3. It should be cleaned up on every boot. It's no problem, we have to add few strings in bootcounter script. @MaxS0niX is in progress...

kar200 commented:
> > I was wondering if there was a way to achieve this ?
> 
> I squashed all partitions after the second kernel into the one big ubi. So there is no problem even the bootflag set up to Sercomm1. U-boot checks rootfs2, gets wrong CRC and switches to Sercomm0 with OpenWrt. On your device (and turbo+ too) there is a bootcounter in mtd3. It should be cleaned up on every boot. It's no problem, we have to add few strings in bootcounter script. @MaxS0niX is in progress...

On the S3 there is no boot counter (nor there is a message about it at boot). It is replaces by some "firmware checksum" that I didn't bother looking into as it doesn't really do much.
The S3 will just try and boot the current kernel - if fail switch to alternative - if fail goes into the recovery mode. Which by the way I have not been able to understand. I tried with wireshark and not a single packet comes out from the router in that mode.

I thought it would have been very useful to have a similar option with these devices as with stock. As you would be able to flash from any partition and the router would work correctly. You could  then just boot to the other one and get back to stock.

csharper2005 commented:
> The S3 will just try and boot the current kernel - if fail switch to alternative - if fail goes into the recovery mode.

Giga has the same behaviour.

> I thought it would have been very useful to have a similar option with these devices as with stock. 

That is definitely so. Dual boot is very useful. But I've chosen big ubi. Rootfs2=24mbytes is too small. In addition, I noticed that the stock firmware is trying to synchronize slot 1 and 2. I did not research this issue more deeply. Do you have any plans to add support for S3? You can implement dual boot. :) It's not a problem to change bootflag from openwrt.

arrmo commented:
> Sorry @arrmo I tagged you here for a different reason (I got confused about the issue).

NP! That makes sense, was wondering why I was getting this one ... LOL!

> Althoguh I don't understand completely I believe it changes the name of the active rootfs during boot depending on what the CMD line from uboot is.

Yes, correct => selects the correct kernel based on an environment variable, and then the rootfs based on some command line handling. See this commit, https://github.com/openwrt/openwrt/pull/4110/commits/62d28f6942e34ce2367ed8c6f9ca4ae315496181

csharper2005 commented:
https://github.com/openwrt/openwrt/pull/4195/commits/47f9ec0dca394f31515152398462205c7229097e - cosmetic fixes, changed baudrate to 57600 (U-Boot default).

kar200 commented:
> Fortunately, this Sercomm device has no boot counters. Newertheless, I found the appropriate command, that can be used for such cases:

Or something simpler (for me anyway). Just split the bootflag partition into 2 part and format the mtd at each boot (there is a bootcount script already that can be used).

````
		partition@300000 {
			label = "bootflag";
			reg = <0x300000 0x40000>;
		};

		partition@340000 {
			label = "bootcount";
			reg = <0x340000 0xC0000>;
		};
````


> I squashed all partitions after the second kernel into the one big ubi

This is actually the best approach for these dual boot devices. And you get a decent amount of space that can be reclaimed by the official firmware if need be.

MaxS0niX commented:
> Or something simpler (for me anyway). Just split the bootflag partition into 2 part and format the mtd at each boot (there is a bootcount script already that can be used).

Or you can: 
Reset counter
boot_count1 - `echo -ne "$(echo FF | sed 's/../\\x&/g')" | dd of=/dev/mtdblock3 bs=1 seek=$((0x20001)) count=1`
boot_count2 - `echo -ne "$(echo FF | sed 's/../\\x&/g')" | dd of=/dev/mtdblock3 bs=1 seek=$((0x20002)) count=1`
Set boot_flag
Sercomm0 - `echo -ne "$(echo 30 | sed 's/../\\x&/g')" | dd of=/dev/mtdblock3 bs=1 seek=7 count=1`
Sercomm1 - `echo -ne "$(echo 31 | sed 's/../\\x&/g')" | dd of=/dev/mtdblock3 bs=1 seek=7 count=1`
or hard option:
`mtd erase /dev/mtd3` or `mtd erase dual-flag`
As a result, both counters are reset and the flag is set "Sercomm0"

csharper2005 commented:
> ... now that you had another reminder to read it when opening the PR...

Fixed. 

> ... and still with a commit title not fulfilling the rules in submitting-patches, now that you had another reminder to read it when opening the PR...

Fixed.

csharper2005 commented:
> Oh please, why this double-posting all the time?

New users usually have not enough experience with git. Me too.

csharper2005 commented:
> Hmm, replacing bootloader is typically a no-go (at least on this target) ...

U-boot is on its place. Now i don't replace it. Is it "go"? :)

csharper2005 commented:
> Hmm, replacing bootloader is typically a no-go (at least on this target) ...

U-boot is on its place. Now i don't replace it. Is it "go"? :)

csharper2005 commented:
> Hmm, replacing bootloader is typically a no-go (at least on this target) ...

U-boot is on its place. Now i don't replace it.

adschm commented on target/linux/ramips/image/Makefile:
```diff
@@ -56,6 +57,22 @@ define Build/loader-kernel
+		PKG_BUILD_DIR="$@.src" \
+		TARGET_DIR="$(dir $@)" LOADER_NAME="$(notdir $@)" \
+		BOARD="$(BOARDNAME)" PLATFORM="$(LOADER_PLATFORM)" \
+		LZMA_TEXT_START=0x82800000 LOADADDR=0x80001000 \
```
What's wrong with using the variable with LOADADDR ...?

csharper2005 replied:
I'll try to explain. We need LOADADDR=0x80001000 here and KERNEL_LOADADDR := 0x81001000 in all other places. Otherwise the devices hangs on boot. I tried to do something but on a not elegant way I think. For example few strings are longer than 73 chars. So I need your advice.

adschm replied:
KERNEL_LOADADDR is actually a device-variable, at least it's defined that way.

Thus, you could simply put "KERNEL_LOADADDR := 0x81001000" into the _device_ definition block, where it would be overwritten just for this device, and leave it untouched for all others.

csharper2005 replied:
Ok, I found another yet option - in current commit.  

The essense if the problem: is if we simple put "KERNEL_LOADADDR := 0x81001000" into the device definition block then this value is used in all device recipes below because they have no own KERNEL_LOADADDR :=  & LZMA_TEXT_START := assignment. I tested that. Therefore KERNEL_LOADADDR :=  & LZMA_TEXT_START := should be assigned in every device recipe that use Build/loader-kernel.

So I made the following changes:
1. Defined device variables KERNEL_LOADADDR & LZMA_TEXT_START;
2. They are assigned for every device via Device/Default recipe and might be overriden in the device recipe.

It works, we have correct load addressess (and lzma text starts too). Tested on this device and Xiaomi AC2100:
binwalk --signature openwrt-ramips-mt7621-beeline_smartbox-giga-squashfs-kernel0.bin
```
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
256           0x100           uImage header, header size: 64 bytes, header CRC: 0xB3629FFA, created: 2021-07-03 11:55:58, image size: 2518229 bytes, Data Address: 0x81001000, Entry Point: 0x81001000, data CRC: 0xD2D58AD2, OS: Linux, CPU: MIPS, image type: OS Kernel Image, compression type: lzma, image name: "MIPS OpenWrt Linux-5.4.128"
320           0x140           LZMA compressed data, properties: 0x6D, dictionary size: 8388608 bytes, uncompressed size: 2486881 bytes
```
binwalk --signature openwrt-ramips-mt7621-xiaomi_redmi-router-ac2100-squashfs-kernel1.bin
```
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             uImage header, header size: 64 bytes, header CRC: 0x7DF37C72, created: 2021-07-03 11:55:58, image size: 2486868 bytes, Data Address: 0x80001000, Entry Point: 0x80001000, data CRC: 0xD86D046F, OS: Linux, CPU: MIPS, image type: OS Kernel Image, compression type: none, image name: "MIPS OpenWrt Linux-5.4.128"
5484          0x156C          Copyright string: "Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>"
5692          0x163C          LZMA compressed data, properties: 0x6D, dictionary size: 8388608 bytes, uncompressed size: 7893429 bytes
```

csharper2005 replied:
@adschm Should I fix anything else?

adschm commented on target/linux/ramips/image/Makefile:
```diff
@@ -56,6 +57,22 @@ define Build/loader-kernel
+	$(MAKE) -C lzma-loader \
+		PKG_BUILD_DIR="$@.src" \
+		TARGET_DIR="$(dir $@)" LOADER_NAME="$(notdir $@)" \
+		BOARD="$(BOARDNAME)" PLATFORM="$(LOADER_PLATFORM)" \
```
$(BOARDNAME) ? Does this still exist for the devices in this target?

csharper2005 replied:
This exists and contains the following: BOARD="**MT7621 based boards**".

adschm replied:
So, it looks like it's not doing anything useful ...

csharper2005 replied:
Would you like me to remove it and test?

adschm replied:
I would keep what's in the original implementation, and I think it's there as well ...

adschm commented on target/linux/ramips/image/Makefile:
```diff
@@ -56,6 +57,22 @@ define Build/loader-kernel
+		PKG_BUILD_DIR="$@.src" \
+		TARGET_DIR="$(dir $@)" LOADER_NAME="$(notdir $@)" \
+		BOARD="$(BOARDNAME)" PLATFORM="$(LOADER_PLATFORM)" \
+		LZMA_TEXT_START=0x82800000 LOADADDR=0x80001000 \
```
So, the only difference to the existing recipe is the TEXT_START value? This should be made a variable instead of copying the whole recipe (don't we already have a solution for that?) ..

csharper2005 replied:
I have done this. It works, but looks a little bit strange. I'm ready to improve this, but I need a piece of advice.  

P.S. When I added KERNEL_LOADADDR := 0x81001000 (and LZMA_TEXT_SIZE too) to the this device definition in mt7621.mk then it has broken another devices (tested on Xiaomi RM2100). KERNEL_LOADADDR := 0x81001000 applied to the Xiaomi too. Currently I don't understand why.

csharper2005 replied:
This conversation is no longer relevant due to code changes.

adschm commented on target/linux/ramips/image/Makefile:
```diff
@@ -117,6 +134,12 @@ define Build/sercom-footer
 	$(call Build/sercom-seal,-f)
 endef
 
+define Build/sercomm-kernel
```
Common spelling seams to be with one "m" ...?

csharper2005 replied:
Common spelling is with double m:
1. OpenWrt code - https://github.com/openwrt/openwrt/search?q=sercomm
2. Company name - https://www.sercomm.com/contpage.aspx?langid=1&type=info&L1id=1&L2id=1

Nevertheless, if I need to fix double m, please, let me know your decision.

csharper2005 commented on target/linux/ramips/dts/mt7621_beeline_smartbox-giga.dts:
```diff
@@ -0,0 +1,164 @@
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x0000>;
+		mtd-mac-address = <&factory 0x21000>;
```
0x4 contains something, but not very useful for DTS. 
`0000000000: 03 76 00 02 B0 39 56 22 │ 24 08 03 76 C3 14 FF FF  ♥v ☻°9V"$◘♥vГ¶яя`
Mac 2g on stock - 60:ce:*:*:*:1a

adschm replied:
Okay, then it's fine the way you did it originally.

adschm commented:
> @adschm,
> I see. Sorry for that. But your comment has given me an insight. I've read the rules many times and assumed that they are related to a first post of the conversation at github (not to a commit message). All is fixed for now. What next steps are necessary from my side?
> 
> P.S. I've looked at a lot of PR's: common mistakes, waste of resources (both requesters and maintainers). Let's add Top10 mistakes to a FAQ. And maybe few words about the OpenWRT change management process. How do you like this idea?

FYI: After a lot of suffering elsewhere ;-), I've followed your suggestion and created a common mistakes page here: https://openwrt.org/docs/guide-developer/frequent-pr-mistakes

It might need (and will receive) some tuning, but it's a start.

csharper2005 commented:
@adschm, great news!!! 

P.S. I sure that 1st level support staff who can filter the 80-90% common mistakes and give link to the wiki post would be very helpful. :) It is not so good when developers spend their time on trivial things. It will alow you concentrate on really important and difficult topics.

csharper2005 commented on scripts/sercomm-kernel.sh:
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env bash
+#
```
Fixed.

csharper2005 commented on target/linux/ramips/dts/mt7621_beeline_smartbox-giga.dts:
```diff
@@ -0,0 +1,164 @@
+
+	chosen {
+		bootargs = "console=ttyS0,57600n8";
+	};
```
You're right. Fixed.

csharper2005 commented:
Images was tested on:
this device - both initramfs & sysupgrade
xiaomi redmi ac2100 - initramfs

adschm commented on target/linux/ramips/image/Makefile:
```diff
@@ -11,20 +11,33 @@ DEVICE_VARS += BUFFALO_TAG_PLATFORM BUFFALO_TAG_VERSION BUFFALO_TAG_MINOR
 loadaddr-$(CONFIG_TARGET_ramips_rt288x) := 0x88000000
 loadaddr-$(CONFIG_TARGET_ramips_mt7621) := 0x80001000
+loadaddr-$(CONFIG_TARGET_ramips_mt7621_DEVICE_beeline_smartbox-giga) := \
+	0x81001000
```
This does not work. This is only evaluated once per (sub)target, so if you built multiple devices, all will have the same value ...
Same for the change below.

csharper2005 replied:
Ok. Fixed.

csharper2005 commented:
https://github.com/openwrt/openwrt/commit/e427a681a8a7ec849175a77f6c2b5fe8f3b0369d - Leds behavouir was made closer to stock

csharper2005 commented:
https://github.com/openwrt/openwrt/pull/4195/commits/bbb5e3cf02f506e8f1fc8d888bf6ea99857a5954 - rebase w/current master

csharper2005 commented:
https://github.com/openwrt/openwrt/pull/4195/commits/f0878ad2eb4f21da90f77bfea6bee03d5473c893 - mtd-mac-addresses -> nvmems

csharper2005 commented:
> Hmm, replacing bootloader is typically a no-go (at least on this target) ...

ping6 -s 1500 @adschm. The commit has been adjusted and bootloader replacing is no longer required. Could you help with review?

adschm commented on target/linux/ramips/dts/mt7621_beeline_smartbox-giga.dts:
```diff
@@ -0,0 +1,184 @@
+	nvmem-cells = <&macaddr_factory_21000>;
+	nvmem-cell-names = "mac-address";
+	mac-address-increment = <1>;
+};
```
This can be put directly into the node.

csharper2005 replied:
Done. The code has been moved into the wan port node.

adschm commented on target/linux/ramips/image/Makefile:
```diff
@@ -5,12 +5,14 @@
 include $(INCLUDE_DIR)/image.mk
 
-DEVICE_VARS += LOADER_TYPE LOADER_FLASH_OFFS
+DEVICE_VARS += KERNEL_LOADADDR LOADER_TYPE LOADER_FLASH_OFFS
```
KERNEL_LOADADDR is already in DEFAULT_DEVICE_VARS, so does not need to be specified here again.

csharper2005 replied:
Done. Thanks.

adschm commented on target/linux/ramips/image/Makefile:
```diff
@@ -159,6 +170,8 @@ endef
 define Device/Default
   PROFILES = Default
+  KERNEL_LOADADDR := $(loadaddr-y)
+  LZMA_TEXT_START := $(lzmatextstart-y)
```
If I've checked correctly, you don't need lzmatextstart-y here at all, but could just put the literal value here.

csharper2005 replied:
Yes, we can just put the literal value here without using lzmatextstart-y. I've removed lzmatextstart-y. Done.

I have made all the changes that you have requested. FW test on a real HW was successful. Many thanks for the review!

csharper2005 commented:
https://github.com/openwrt/openwrt/pull/4195/commits/2a80a5b0592048bbfb9eee81940a5621bad27bb4 - removed kmod-mt7663-firmware-sta, see https://github.com/openwrt/mt76/issues/572

csharper2005 commented:
https://github.com/openwrt/openwrt/pull/4195/commits/b1250ac29c302967a9053fb05d0845f7f35d9ae6 - common with other sercomm devices functions moved into a separate (this) commit
https://github.com/openwrt/openwrt/pull/4195/commits/e59b3ee0a56c3c84a48010c12921e58ce9dc6166 - patch to avoid [FS#3887](https://bugs.openwrt.org/index.php?do=details&task_id=3887) and make installation safe on stock U-Boot

adschm commented:
Can this parser be upstreamed?

csharper2005 commented:
https://github.com/openwrt/openwrt/commit/0d96d142914908be9e865d20ea40299549caa1aa - Added recipe for factory image creation. Updated OpenWrt installation instruction (easy installation via WEB interface).

csharper2005 commented:
> Can this parser be upstreamed?

If I understand correctly we are talking about Linux. To be honest I have a lack of experience to asses correctly if it really necessary and how to do that. Could you please explain if upstreaming is only an idea or a jobstopper for the PR approve from your or OpenWrt side?

adschm commented:
> > Can this parser be upstreamed?
> 
> If I understand correctly we are talking about Linux. To be honest I have a lack of experience to asses correctly if it really necessary and how to do that. Could you please explain if upstreaming is only an idea or a jobstopper for the PR approve from your or OpenWrt side?

Well, essentially you need somebody to review that parser. So far, nobody has shown up who seems to be interested to do that here. So maybe, upstream will have more people capable and/or interested in doing so.

Backporting from kernel is much easier to accept here then, since we reviewer can expect that somebody competent has checked whether the parser is both reasonable and working.

This will of course also move the maintenance burden to those who are most capable of doing it...

MaxS0niX commented:
> > 
> 
> Well, essentially you need somebody to review that parser. So far, nobody has shown up who seems to be interested to do that here. So maybe, upstream will have more people capable and/or interested in doing so.
> 
> Backporting from kernel is much easier to accept here then, since reviewers can expect that somebody competent has checked whether the parser is both reasonable and working.
> 
> This will of course also move the maintenance burden to those who are most capable of doing it...

This patch is required for the stock SerComm U-boot bootloader to work. Can an alternative bootloader be used? For example U-boot from SDK MTK or do we have to officially port to U-boot Denx MT7621?

While we are looking for a reviewer for this patch...

csharper2005 commented:
> > > Can this parser be upstreamed?
> > 
> > 
> > If I understand correctly we are talking about Linux. To be honest I have a lack of experience to asses correctly if it really necessary and how to do that. Could you please explain if upstreaming is only an idea or a jobstopper for the PR approve from your or OpenWrt side?
> 
> Well, essentially you need somebody to review that parser. So far, nobody has shown up who seems to be interested to do that here. So maybe, upstream will have more people capable and/or interested in doing so.
> 
> Backporting from kernel is much easier to accept here then, since reviewers can expect that somebody competent has checked whether the parser is both reasonable and working.
> 
> This will of course also move the maintenance burden to those who are most capable of doing it...

Ok. I see. We'll try to find reviewer. While the reviewer hasn't been found yet... What do you think if I remove this patch from the Pr? Will you accept it? We get the situation like with Netgears (which are already in OpenWrt) in this case. Something better than nothing.

We add the patch in a separate pr when it will be in upstream.

csharper2005 commented:
https://github.com/openwrt/openwrt/pull/4195/commits/ac56cc01d498c07a6dc8d168b2c72ffb9b34b439 - removed unnecessary rootfs and kernel images. installation instruction was made simplier
https://github.com/openwrt/openwrt/pull/4195/commits/081229efac5aef9e8b19607014a784dd7e3ebc9e - sc partition patch adaptation for 5.10 kernel

csharper2005 commented:
[abf6b52](https://github.com/openwrt/openwrt/commit/abf6b52dbf0f222dbef2036b25b69553d21adfe6) - merge conflict was resolved

csharper2005 commented:
[c49a8dc](https://github.com/openwrt/openwrt/pull/4195/commits/c49a8dc6c8d7fc128e426cc52eb976ce34cd436e) - fixed gpio groups in dts

csharper2005 committed:
scripts: add support for Sercomm kernel header

This scripts creates custom kernel header that necessary for Sercomm
mt7621 devices:
- Sercomm S3
- Beeline SmartBox Giga
- Beeline SmartBox Pro
- Beeline Smartbox Turbo
- Beeline Smartbox Turbo+
- WiFire S1500.NBN

Header format
-------------
+--------+---------------+------------------------+
| Offset | Value         | Description            |
+========+===============+========================+
| 0x0    | 53 65 72 00   | Magic "Ser."           |
+--------+---------------+------------------------+
| 0x4    | 04 00 00 01   | End offset of RootFS   |
+--------+---------------+------------------------+
|        |               | This header checksum   |
| 0x8    | d6 14 9a c1   | htonl(~crc)            |
+--------+---------------+------------------------+
| 0xc    | 02 ff ff ff   | Constant               |
+--------+---------------+------------------------+
| 0x10   | 00 01 40 00   | Kernel start offset    |
+--------+---------------+------------------------+
| 0x14   | c6 94 24 00   | Kernel length          |
+--------+---------------+------------------------+
|        |               | Kernel checksum        |
| 0x18   | e7 78 89 f1   | htonl(~crc)            |
+--------+---------------+------------------------+
| 0x1c   | 00 00 00 00   | Constant               |
+--------+---------------+------------------------+
| 0x20   | ff ff ff ff   | Constant               |
+--------+---------------+------------------------+
| 0x24   | ff ff ff ff   | Constant               |
+--------+---------------+------------------------+
| 0x28   | 00 00 00 01   | RootFS offset          |
+--------+---------------+------------------------+
|        |               | RootFS length          |
| 0x2c   | 04 00 00 00   | Always 0x4, we check   |
|        |               | UBI magic only         |
+--------+---------------+------------------------+
|        |               | RootFS checksum        |
| 0x30   | 1c fc 55 2d   | htonl(~crc)            |
|        |               | Const for UBI magic    |
+--------+---------------+------------------------+
| 0x34   | 00 00 00 00   | Constant               |
+--------+---------------+------------------------+
| 0x38   | ff ff ff ff … | Pad to 0x100           |
+--------+---------------+------------------------+

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>

csharper2005 committed:
ramips: add support for Beeline SmartBox GIGA

Beeline SmartBox GIGA is a wireless WiFi 5 router manufactured by
Sercomm company.

Device specification
--------------------
SoC Type: MediaTek MT7621AT
RAM: 256 MiB, Nanya NT5CC128M16JR-EK
Flash: 128 MiB, Macronix MX30LF1G18AC
Wireless 2.4 GHz (MT7603EN): b/g/n, 2x2
Wireless 5 GHz (MT7613BE): a/n/ac, 2x2
Ethernet: 3 ports - 2xGbE (WAN, LAN1), 1xFE (LAN2)
USB ports: 1xUSB3.0
Button: 1 button (Reset/WPS)
PCB ID: DBE00B-1.6MM
LEDs: 1 RGB LED
Power: 12 VDC, 1.5 A
Connector type: barrel
Bootloader: U-Boot

Installation
-----------------
1. Downgrade stock (Beeline) firmware to v.1.0.02;
2. Give factory OpenWrt image a shorter name, e.g. 1001.img;
3. Upload and update the firmware via the original web interface.

Remark: You might need make the 3rd step twice if your running firmware
is booted from the Slot 1 (Sercomm0 bootflag). The stock firmware
reverses the bootflag (Sercomm0 / Sercomm1) on each firmware update.

Revert to stock
---------------
1. Change the bootflag to Sercomm1 in OpenWrt CLI and then reboot:
      printf 1 | dd bs=1 seek=7 count=1 of=/dev/mtdblock3
2. Optional: Update with any stock (Beeline) firmware if you want to
   overwrite OpenWrt in Slot 0 completely.

MAC Addresses
-------------
+-----+-----------+---------+
| use | address   | example |
+-----+-----------+---------+
| LAN | label     | *:16    |
| WAN | label + 1 | *:17    |
| 2g  | label + 4 | *:1a    |
| 5g  | label + 5 | *:1b    |
+-----+-----------+---------+
The label MAC address was found in Factory 0x21000

Notes
-----
1. The following scripts are required for the build:
      sercomm-crypto.py - already exists in OpenWrt
      sercomm-partition-tag.py - already exists in OpenWrt
      sercomm-payload.py - already exists in OpenWrt
      sercomm-pid.py - new, the part of this pull request
      sercomm-kernel-header.py - new, the part of this pull request
2. This device (same as other Sercomm S2,S3-based devices) requires
   special LZMA and LOADADDR settings for successful boot:
      LZMA_TEXT_START=0x82800000
      KERNEL_LOADADDR=0x81001000
      LOADADDR=0x80001000
3. This device (same as several other Sercomm-based devices - Beeline,
   Netgear, Etisalat, Rostelecom) has partition map (mtd1) containing
   real partition offsets, which may differ from device to device
   depending on the number and location of bad blocks on NAND.
   "fixed-partitions" is used if the partition map is not found or
   corrupted. This behavour (it's the same as on stock firmware) is
   provided by MTD_SERCOMM_PARTS module.

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>

csharper2005 committed:
scripts: add support for Sercomm PID

This scripts creates Sercomm PID file. PID is necessary for the factory
images creation of variuos Sercomm-based devices (Beeline, Netgear,
Etisalat).

Size: 0x70
+-------+------+---------------+------------------+
| Start | Size | Value* (ASCII)| Description      |
+=======+======+===============+==================+
| 0x0   | 0x8  | 10100         | Hardware version |
+-------+------+---------------+------------------+
| 0x8   | 0x8  | 444245 (DBE)  | Hardware ID      |
+-------+------+---------------+------------------+
| 0x64  | 0x4  | 1002          | Software version |
+-------+------+---------------+------------------+
*for Beeline Smartbox GIGA

Signed-off-by: Mikhail Zhilkin <csharper2005@gmail.com>

csharper2005 commented:
@adschm @hauke @mans0n
Hi guys! 

I and would like to to dust off the PR and ask you for your advice.

Known blocker for the PR:
- [ ] Reviewer not found for the commit [kernel: add partition table parser for Sercomm partition map](https://github.com/openwrt/openwrt/pull/4195/commits/30e313f69f140c3da7bf3f408efd5d0319dd486a)

### Possible options
**Option 1.** Remove the blocking commit
- [+] The simpliest way. We don't need a reviewer and upstreaming.
- [-] Bootlops after the OpenWrt installation (e.g. FS#3887) and EEPROM displacement (e.g. FS#2886, FS#3582) are possible on devices with bad blocks. Workaround - Users are forced to use alternative bootloaders (e.g. breed) or / and move EEPROM. 
- [-] Affected devices - all devices with Sercomm partition map, e.g.:
  - NetGear R6260 (already in OpenWrt)
  - NetGear R6350 (already in OpenWrt)
  - NetGear R6850 (already in OpenWrt)
  - NetGear WAC124 (CHJ) (already in OpenWrt)
  - NetGear R6800 (already in OpenWrt)
  - NetGear R6700-v2 (already in OpenWrt)
  - NetGear R7200 (already in OpenWrt)
  - NetGear Nighthawk AC2400 (BZV)  (already in OpenWrt)
  - Beeline SmartBox Turbo+ (PR is in place)
  - Beeline SmartBox Turbo (under development)
  - Beeline SmartBox GIGA  (this PR)
  - Sercomm S3 (under development) 

**Option 2.** Merge the commit with Sercomm mtd parser
- [+] Bugs (e.g. FS#2886, FS#3887, FS#3582) can be resolved. Including Netgear devices those are already commited to OpenWrt. 
- [+] Simple and clear installation guide via OEM web interface. Bootloader replacement is not necessary.
- [+] Patch was successfully tested on dozens and maybe even hundreds Beeline Smartbox GIGA and Turbo+ devices. And there is no negative impact on other ramips devices built from the same repository (e.g. Xiaomi AC2100, Newifi 3 D2)
- [-] No reviewer from @openwrt team. Upstreaming to Linux can take months and even years. I have no ideas about the next steps here. 

### Questions
- [ ] Are there any other options / ideas here from your point of view?
- [ ] Is it ok if I choose option 1? Do you notice any additional blockers except commit with Sercomm mtd parser?

Thanks in advance!

rmilecki commented:
Python really doesn't fit well for working with binary formats, it doesn't have structs you could use for storing packed data.

I think `sercomm-kernel-header.py` and `sercomm-pid.py` should really be written as C programs. C structs would result in cleaner code & simplified filling all fields. Without all that hack jumping to hardcoded offsets.

rmilecki commented on target/linux/ramips/files/drivers/mtd/parsers/scpart.c:
```diff
@@ -0,0 +1,243 @@
+		parts[nr_parts].of_node = pp;
+
+		if (!of_property_read_string(pp, "label", &partname) ||
+				!of_property_read_string(pp, "name", &partname))
```
Why do we have two properties for partition name?

csharper2005 replied:
Fixed. `!of_property_read_string(pp, "name", &partname)) ` dropped.

rmilecki commented on target/linux/ramips/files/drivers/mtd/parsers/scpart.c:
```diff
@@ -0,0 +1,243 @@
+
+		/* reallocate partition array */
+		parts_tmp = parts;
+		parts = kzalloc((nr_parts + 1) * sizeof(*parts), GFP_KERNEL);
```
Use `kcalloc`?

csharper2005 replied:
> Use `kcalloc`?

Yes. Fixed.

rmilecki commented on target/linux/ramips/files/drivers/mtd/parsers/scpart.c:
```diff
@@ -0,0 +1,243 @@
+		if (parts_tmp) {
+			memcpy(parts, parts_tmp, nr_parts * sizeof(*parts));
+			kfree(parts_tmp);
+		}
```
First: I think you could use `realloc`. Secondly: use `of_get_child_count()` and alloc big enough array before the `for_each_child_of_node()`.

csharper2005 replied:
Indeed... Done.

rmilecki commented on target/linux/ramips/dts/mt7621_beeline_smartbox-giga.dts:
```diff
@@ -0,0 +1,237 @@
+			reg = <0x7c00000 0x380000>;
+			scpart-id = <11>;
+			read-only;
+		};
```
If I'm not mistaken, your Linux driver ignores all those offsets & sizes. It seems the are also dynamic (stored in struct on flash). What's the point of them in DT then?

csharper2005 replied:
`compatible = "fixed-partitions"` is a fallback strategy if partition map is broken or missing. So, we need those offsets & sizes. And if a device has no bad blocks or bad blocks in some places all will be ok. 

IOW the layout differs from the reference one ("fixed-partitions") only if there are bad blocks on NAND. Partitions are shifted farther on the bad blocks size .

Partition map is ok:
```
Mon Mar 28 04:41:08 2022 kern.info kernel: [    1.096374] scpart: Valid 'SC PART MAP' (12 partitions) found at 0x100000
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.109964] 12 scpart partitions found on MTD device mt7621-nand
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.121917] Creating 12 MTD partitions on "mt7621-nand":
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.132501] 0x000000000000-0x000000100000 : "u-boot"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.153524] 0x000000100000-0x000000200000 : "dynamic partition map"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.177207] 0x000000200000-0x000000300000 : "Factory"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.198437] 0x000000300000-0x000000400000 : "Boot Flag"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.219971] 0x000000400000-0x000000a00000 : "kernel"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.291609] 0x000000a00000-0x000001000000 : "Kernel 2"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.363526] 0x000001000000-0x000002800000 : "ubiconcat0"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.617160] 0x000002800000-0x000004000000 : "File System 2"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.871459] 0x000004000000-0x000004800000 : "ubiconcat1"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    1.963858] 0x000004800000-0x000005400000 : "ubiconcat2"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    2.096453] 0x000005400000-0x000007c00000 : "ubiconcat3"
Mon Mar 28 04:41:08 2022 kern.notice kernel: [    2.511879] 0x000007c00000-0x000007f80000 : "bad block reserved"
```
Partition map missing:
```
[    2.648797] scpart: No valid 'SC PART MAP' 
[    2.657036] 12 fixed-partitions partitions found on MTD device mt7621-nand 
[    2.670726] Creating 12 MTD partitions on "mt7621-nand": 
[    2.681318] 0x000000000000-0x000000100000 : "u-boot" 
[    2.702380] 0x000000100000-0x000000200000 : "dynamic partition map" 
[    2.726119] 0x000000200000-0x000000300000 : "Factory" 
[    2.747394] 0x000000300000-0x000000400000 : "Boot Flag" 
[    2.768901] 0x000000400000-0x000000a00000 : "kernel" 
[    2.840453] 0x000000a00000-0x000001000000 : "Kernel 2" 
[    2.912051] 0x000001000000-0x000002800000 : "ubiconcat0" 
[    3.164790] 0x000002800000-0x000004000000 : "File System 2" 
[    3.418140] 0x000004000000-0x000004800000 : "ubiconcat1" 
[    3.510168] 0x000004800000-0x000005400000 : "ubiconcat2" 
[    3.642439] 0x000005400000-0x000007c00000 : "ubiconcat3" 
[    4.056060] 0x000007c00000-0x000007f80000 : "bad block reserved"
```

csharper2005 replied:
@rmilecki First of all, I would like to thank you for the review and useful feedback. I've started upstreaming before your review, so I need to send the version 2 after the changes. Please take a look, maybe there is something else that needs fixing.

@adschm @hauke @mans0n Can we at least assign a "new device" label, please?

csharper2005 commented on target/linux/ramips/files/drivers/mtd/parsers/scpart.c:
```diff
@@ -0,0 +1,243 @@
+{
+	return !!((pdesc->part_id != 0xFFFFFFFFUL) &&
+			(pdesc->part_offs != 0xFFFFFFFFUL) &&
+			(pdesc->part_bytes != 0xFFFFFFFFUL));
```
Thanks! Fixed.

csharper2005 commented on target/linux/ramips/files/drivers/mtd/parsers/scpart.c:
```diff
@@ -0,0 +1,243 @@
+
+#define	MAP_OFFS_IN_BLK	(0x800)
+
+#define	MAP_MIRROR_NUM	(2)
```
Fixed.

csharper2005 commented on target/linux/ramips/files/drivers/mtd/parsers/scpart.c:
```diff
@@ -0,0 +1,243 @@
+	}
+
+	res2 = mtd_read(master, partmap_offs, master->erasesize, &retlen, buf);
+	if (res2 || (retlen != master->erasesize)) {
```
Extra braces  deleted.

csharper2005 commented on target/linux/ramips/files/drivers/mtd/parsers/scpart.c:
```diff
@@ -0,0 +1,243 @@
+
+out:
+	if (res > 0)
+		pr_info("%s: valid 'SC PART MAP' found (%d partitions)\n", MOD_NAME, res);
```
Done.

csharper2005 commented on target/linux/ramips/files/drivers/mtd/parsers/scpart.c:
```diff
@@ -0,0 +1,243 @@
+
+	offs = 0;
+	while ((magic_found < MAP_MIRROR_NUM) &&
+			(offs < master->size) && !mtd_block_isbad(master, offs)) {
```
Yes. Part map and bootloader are at the start of the NAND and they must be without bad block. If it is still there, it is pointless to seek further. And I've never seen Sercomm routers with bad blocks on mtd0, mtd1. Do you suggest to drop !mtd_block_isbad(master, offs)?

csharper2005 commented:
> I think `sercomm-kernel-header.py` and `sercomm-pid.py` should really be written as C programs. C structs would result in cleaner code & simplified filling all fields. Without all that hack jumping to hardcoded offsets.

Thanks for your idea about the structs! They exist in Python (struct.pack_into) and I managed to get cleaner code & simplified filling now, However, I am willing to rewrite in C if necessary. Please let me know if so.

Note that 3 of 5 py scripts from my recipe (`sercomm-partition-tag.py`, `sercomm-payload.py`, `sercomm-crypto.py`) are already merged in OpenWrt (it was reuse of existing scripts). Another open question is how to commit from this PR to https://git.openwrt.org/?p=project/firmware-utils.git;a=summary Looks like firmware-utils were dropped from github.

Looking forward to your feedback.

csharper2005 commented:
https://github.com/openwrt/openwrt/pull/4195/commits/a0c7149dbb3ba4594bfcdae33e74da63575ca978 
- rebase
- decomposition / split of the large factory recipe
- move recipes to the own .mk
- small improvements of the scripts

csharper2005 commented:
@rmilecki  upstreamed patch was backported. Works well with 5.10 & 5.15 both. 
@mans0n, @hauke: @adschm last seen here a long time and PR still has [needs changes] label. Could you help with further processing of this PR?
```
Thu May 12 18:12:27 2022 kern.info kernel: [    2.171290] scpart: Valid 'SC PART MAP' (12 partitions) found at 0x100000
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.184906] 12 scpart partitions found on MTD device mt7621-nand
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.196867] Creating 12 MTD partitions on "mt7621-nand":
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.207454] 0x000000000000-0x000000100000 : "u-boot"
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.228623] 0x000000100000-0x000000200000 : "dynamic partition map"
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.252274] 0x000000200000-0x000000300000 : "Factory"
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.273647] 0x000000300000-0x000000400000 : "Boot Flag"
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.295285] 0x000000400000-0x000000a00000 : "kernel"
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.366468] 0x000000a00000-0x000001000000 : "Kernel 2"
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.437893] 0x000001000000-0x000002800000 : "File System 1"
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.691258] 0x000002800000-0x000004020000 : "File System 2"
Thu May 12 18:12:27 2022 kern.notice kernel: [    2.945151] 0x000004020000-0x000004820000 : "Configuration/log"
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.038443] 0x000004820000-0x000005420000 : "application tmp buffer (Ftool)"
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.174111] 0x000005420000-0x000007c40000 : "container"
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.588192] 0x000007c40000-0x000007f80000 : "bad block reserved"
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.641428] Concatenating MTD devices:
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.648951] (0): "File System 1"
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.655393] (1): "Configuration/log"
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.662498] (2): "application tmp buffer (Ftool)"
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.671853] (3): "container"
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.677594] into device "ubi-concat"
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.684779] 1 fixed-partitions partitions found on MTD device ubi-concat
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.698126] Creating 1 MTD partitions on "ubi-concat":
Thu May 12 18:12:27 2022 kern.notice kernel: [    3.708362] 0x000000000000-0x000005400000 : "ubi"
```

csharper2005 commented:
Backported patches have been refreshed.

csharper2005 commented:
Backported patches have been refreshed with fix from Yang Yingliang https://lore.kernel.org/linux-mtd/20220617014008.851583-1-yangyingliang@huawei.com/

SopaDeMacaco-UmaDelicia commented:
@hauke, @mans0n, @981213 guys please review this PR, it's been a long year and a plenty of people really want to see this router upstreamed.

csharper2005 commented:
@mans0n I've seen your iptime pr and know that you're experienced in KERNEL load addresses, LZMA. What do you think about the changes in Makefile here?

981213 commented on target/linux/ramips/image/common-sercomm.mk:
```diff
@@ -0,0 +1,87 @@
+  KERNEL_LOADADDR := 0x81001000
+  LZMA_TEXT_START := 0x82800000
+  KERNEL := kernel-bin | append-dtb | lzma | loader-kernel | lzma | \
+	uImage lzma | sercomm-kernel
```
The kernel is compressed twice. Is it possible to skip one of them? Could you try to drop the first lzma and replace loader-kernel with relocate-kernel, or drop the second lzma and use uImage none instead of uImage lzma for the uImage header?

csharper2005 replied:
@981213 glad to see you here! Yeah, I tried it again. 
1st option (dropping the first lzma and replace loader-kernel with relocate-kernel) - it hangs on "Starting kernel":
```
## Booting image at bc400100 ...
   Image Name:   MIPS OpenWrt Linux-5.15.49
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    2907229 Bytes =  2.8 MB
   Load Address: 81001000
   Entry Point:  81001000
   Verifying Checksum ... OK
   Uncompressing Kernel Image ... OK
kernel addr :0xBC400100
No initrd
## Transferring control to Linux (at address 81001000) ...
## Giving linux memsize in MB, 256

Starting kernel ...
```
2nd option (dropping the second lzma and use uImage none instead of uImage lzma for the uImage header):
```
## Booting image at bc400100 ...
   Image Name:   MIPS OpenWrt Linux-5.15.49
   Image Type:   MIPS Linux Kernel Image (uncompressed)
   Data Size:    2868290 Bytes =  2.7 MB
   Load Address: 81001000
   Entry Point:  81001000
   Verifying Checksum ... OK
Unimplemented compression type 0
kernel check fail
```
We can use fast compression for the second lzma (-a0). What do you think? 

Successful boot (current recipe from the commit):
```
*************************************
Boot Flag : Sercomm0
*************************************
Fw header Magic check OK!
FW header checksum should: 0x7ec29420, crc32 result: 0x7ec29420, Fw header CRC c                                                                                                             heck OK!
kernel : real offset: 0x00400100, data length: 0x002c5c9d,checksum :0xe93a226b
kernel result: 0xe93a226b, kernel check sum ok!
rootfs:  real offset: 0x01000000, data length: 0x00000004, checksum :0x2d55fc1c
rootfs result: 0x2d55fc1c, rootfs check sum ok!
Kernel Addr : 0xbc400100
## Booting image at bc400100 ...
   Image Name:   MIPS OpenWrt Linux-5.15.49
   Image Type:   MIPS Linux Kernel Image (lzma compressed)
   Data Size:    2907229 Bytes =  2.8 MB
   Load Address: 81001000
   Entry Point:  81001000
   Verifying Checksum ... OK
   Uncompressing Kernel Image ... OK
kernel addr :0xBC400100
No initrd
## Transferring control to Linux (at address 81001000) ...
## Giving linux memsize in MB, 256

Starting kernel ...



OpenWrt kernel loader for MIPS based SoC
Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>
Decompressing kernel... done!
Starting kernel at 80001000...
```

981213 replied:
> We can use fast compression for the second lzma (-a0). What do you think?

OK. Let's go with that for now.

csharper2005 replied:
> > We can use fast compression for the second lzma (-a0). What do you think?
> 
> OK. Let's go with that.

Thanks, done. Tests are ok.

csharper2005 commented:
https://github.com/openwrt/openwrt/pull/4195/commits/e1063246b3bf71e8ea39871cc0184b2493ac477b - rebased with current master, add `2>/dev/null` to a dd command in Build/sercomm-prepend-tagged-kernel recipe.

hauke commented:
Thank you for the patches, I applied the to master in commit 87092b3c065978fc0e673856c34b3db0829280db, 0cb1dc02771ce1c4cad20017a6d5616718db3877 and bd783fd60a5f9513aa405437efff55fe29cd89c2.
